### ISSUE TITLE
Feature Request: Adding native support for gradient accumulation

### ISSUE THREAD

**Issue by yuchaoran2011** [ml-explore/mlx#2840]
Gradient accumulation is possible today but it has a major issue. When implementing gradient accumulation manually using `tree_map` to accumulate gradients across many steps (e.g., 1000+), memory usage grows linearly with the number of accumulation steps, even though computation memory should only be proportional to batch size. This is likely due to Python object creation overhead from `tree_map`. This explosion in memory usage has repeatedly caused a transformer training loop on my M3 Max to crash the whole machine.

* Problem Description
MLX's `value_and_grad()` computes gradients from scratch each time and returns them as a new tree structure. Unlike PyTorch, which accumulates gradients in-place as `param.grad` attributes, MLX requires manual accumulation using `tree_map`. This creates new Python tree structures (dicts/lists) for each accumulation step, leading to memory growth.

* Reproducing the issue
```python
import mlx.core as mx
import mlx.nn as nn
from mlx.utils import tree_map

model = nn.Linear(1000, 1000)
loss_fn = nn.losses.mse_loss
loss_and_grad_fn = nn.value_and_grad(model, loss_fn)

# Simulate gradient accumulation over many steps
grad_accum_steps = 4096  # Large number of accumulation steps
grads_accum = None

for i in range(grad_accum_steps):
    x = mx.random.normal((16, 1000))
    y = mx.random.normal((16, 1000))
    
    loss, grads = loss_and_grad_fn(model, x, y)
    mx.eval(grads)  # Materialize gradients
    
    if grads_accum is None:
        grads_accum = grads
    else:
        # This creates a new Python tree structure each time
        grads_accum = tree_map(lambda a, b: a + b, grads_accum, grads)
        mx.eval(grads_accum)  # Materialize accumulated gradients
        del grads  # Try to help GC
    
    # Memory usage grows with i, even though computation memory should be constant
```
Expected behavior is that memory usage should remain constant regardless of `grad_accum_steps`.

Adding a built-in mechanism for gradient accumulation would solve this issue. For example, if we can have an in-place version of tree_map, memory efficiency will improve

**Comment by awni** [comment]
The example you posted doesn't run. So I'm not entirely sure what you are running into. It has a couple problems:

- The `loss_and_grad_fn` does not take a `model` as input. It's just the mse_loss which only takes `x` and `y`
- You are not taking a gradient with respect to anything here since the model is not used in the loss function


Modified like so works fine for me:


```python
import mlx.core as mx
import mlx.nn as nn
from mlx.utils import tree_map

model = nn.Linear(1000, 1000)

def loss_fn(x, y):
    return nn.losses.mse_loss(model(x), y)

loss_and_grad_fn = nn.value_and_grad(model, loss_fn)

# Simulate gradient accumulation over many steps
grad_accum_steps = 4096  # Large number of accumulation steps
grads_accum = None

for i in range(grad_accum_steps):
    x = mx.random.normal((16, 1000))
    y = mx.random.normal((16, 1000))
    
    loss, grads = loss_and_grad_fn(x, y)
    mx.eval(grads)  # Materialize gradients
    
    if grads_accum is None:
        grads_accum = grads
    else:
        # This creates a new Python tree structure each time
        grads_accum = tree_map(lambda a, b: a + b, grads_accum, grads)
        mx.eval(grads_accum)  # Materialize accumulated gradients
        del grads  # Try to help GC
    
    # Memory usage grows with i, even though computation memory should be constant
```

So I will close this.

**Comment by awni** [comment]
Actually you're example exposed a memory leak in a certain degenerate case when a function transformation updates a container in a circular fashion without using the input of the container. I will open a separate issue on that.

**Comment by yuchaoran2011** [comment]
@awni Thanks for the quick reply. I tried to provide a simplifying example but that might have created confusion. My loss_fn looks like this:
```
def loss_fn(x, y):
    y = y.reshape(-1)
    logits = model(x)
    logits = logits.reshape(-1, model.config.vocab_size)
    return nn.losses.cross_entropy(logits, y, reduction="mean")
```
so model is actually used in the loss function. I think my point that tree_map creates new structure that consumes memory is still valid. Did I miss something?

**Comment by awni** [comment]
If you can share a full repro that grows memory linearly I’d be happy to take a look

**Comment by yuchaoran2011** [comment]
Sure thank you, let me clean up my code and push to a repo later today and share here

**Comment by yuchaoran2011** [comment]
I experimented a bit more and my theory above should be invalid. Most likely what happened was I missed adding some `mx.eval()` calls during my gradient accumulation loop, causing the computation graph's memory usage to explode. Now with proper explicit materialization, my code no longer crashes the laptop. Here's the code that I just pushed: https://github.com/yuchaoran2011/gpt2-mlx. 

> you're example exposed a memory leak in a certain degenerate case when a function transformation updates a container in a circular fashion without using the input of the container.

Glad to hear that this thread at least yielded some meaningful discovery. I'd love to understand better the nature of that issue and possibly contribute to fixing it.
